google/cloud/firestore_v1/base_query.py

```diff
@@ -315,36 +422,66 @@ def where(self, field_path: str, op_string: str, value) -> "BaseQuery":
             modified with the newly added filter.
 
         Raises:
-            ValueError: If ``field_path`` is invalid.
-            ValueError: If ``value`` is a NaN or :data:`None` and
-                ``op_string`` is not ``==``.
+            ValueError: If
+                * ``field_path`` is invalid.
+                * If ``value`` is a NaN or :data:`None` and ``op_string`` is not ``==``.
+                * FieldFilter was passed without using the filter keyword argument.
+                * `And` or `Or` was passed without using the filter keyword argument .
+                * Both the positional arguments and the keyword argument `filter` were passed.
         """
-        field_path_module.split_field_path(field_path)  # raises
 
-        if value is None:
-            if op_string != _EQ_OP:
-                raise ValueError(_BAD_OP_NAN_NULL)
-            filter_pb = query.StructuredQuery.UnaryFilter(
-                field=query.StructuredQuery.FieldReference(field_path=field_path),
-                op=StructuredQuery.UnaryFilter.Operator.IS_NULL,
+        if isinstance(field_path, FieldFilter):
+            raise ValueError(
+                "FieldFilter object must be passed using keyword argument 'filter'"
             )
-        elif _isnan(value):
-            if op_string != _EQ_OP:
-                raise ValueError(_BAD_OP_NAN_NULL)
-            filter_pb = query.StructuredQuery.UnaryFilter(
-                field=query.StructuredQuery.FieldReference(field_path=field_path),
-                op=StructuredQuery.UnaryFilter.Operator.IS_NAN,
+        if isinstance(field_path, BaseCompositeFilter):
+            raise ValueError(
+                "'Or' and 'And' objects must be passed using keyword argument 'filter'"
             )
-        elif isinstance(value, (transforms.Sentinel, transforms._ValueList)):
-            raise ValueError(_INVALID_WHERE_TRANSFORM)
-        else:
-            filter_pb = query.StructuredQuery.FieldFilter(
-                field=query.StructuredQuery.FieldReference(field_path=field_path),
-                op=_enum_from_op_string(op_string),
-                value=_helpers.encode_value(value),
+
+        field_path_module.split_field_path(field_path)  # raises
+        new_filters = self._field_filters
+
+        if field_path is not None and op_string is not None:
+            if filter is not None:
+                raise ValueError(
+                    "Can't pass in both the positional arguments and 'filter' at the same time"
+                )
+            warnings.warn(
+                "Detected filter using positional arguments. Prefer using the 'filter' keyword argument instead.",
+                UserWarning,
+                stacklevel=2,
```

Comment:
I'm not sure I understand stacklevel as it's used here - doesn't that mean that "where" will not be displayed as the origin of the warning?

Reply:
When I ran the code, with stacklevel 2, it shows the `where` as the source of issue.

Output using `stacklevel=2`

> /Users/mariattaw/google-github/python-firestore/venv/lib/python3.11/site-packages/google/cloud/firestore_v1/base_collection.py:290: UserWarning: Detected filter using positional arguments. Prefer using the 'filter' keyword argument instead.
  return query.where(field_path, op_string, value)


When I use stacklevel = 1, it shows the `warnings.warn` as the source of issue

Output of using stacklevel = 1
> /Users/mariattaw/google-github/python-firestore/venv/lib/python3.11/site-packages/google/cloud/firestore_v1/base_query.py:450: UserWarning: Detected filter using positional arguments. Prefer using the 'filter' keyword argument instead.
  warnings.warn(


I think it's clearer to show the warning with stacklevel 2 so user knows that `where` has a different signature now. What do you think?

Reply:
Interesting, so the "where" is in the headline of the stack trace. That does make sense. I'm okay with either and defer to your judgement. Thanks for your response.